src/analyses/threadId.ml

```diff
@@ -87,13 +87,25 @@ struct
     else
       None
 
+  (** get the node that identifies the current context, possibly that of a wrapper function *)
+  let indexed_node_for_ctx ?(previous = false) ctx =
+    match ctx.ask (Queries.ThreadCreateIndexedNode previous) with
+    | `Lifted node, `Lifted count -> node, Some count
+    | `Lifted node, `Bot -> node, Some 0
+    | `Lifted node, _ -> node, None
+    | _ -> ctx.prev_node, None
+
   let threadenter ctx lval f args =
-    let+ tid = create_tid ctx.local ctx.prev_node f in
+    (* [ctx] here is the same as in [special], i.e. before incrementing the unique-counter,
+       thus we want the current counter (previous: false) *)
+    let+ tid = create_tid ctx.local (indexed_node_for_ctx ctx) f in
     (tid, TD.bot ())
 
   let threadspawn ctx lval f args fctx =
     let (current, td) = ctx.local in
-    (current, Thread.threadspawn td ctx.prev_node f)
+    (* here we see the updated counter, so we want the previous counter value *)
+    let node, index = indexed_node_for_ctx ~previous:true ctx in
```

Comment:
Just do the query in `fctx` instead of `ctx`. Then it's querying the counter from the created thread, not the creator thread.